vcd/resource_vcd_vapp_vm.go

```diff
@@ -1365,8 +1364,12 @@ func genericVcdVAppVmRead(d *schema.ResourceData, meta interface{}, origin strin
 		return fmt.Errorf("error storing customzation block: %s", err)
 	}
 
-	_ = d.Set("hardware_version", vm.VM.VmSpecSection.HardwareVersion.Value)
-	_ = d.Set("os_type", vm.VM.VmSpecSection.OsType)
+	if vm.VM.VmSpecSection != nil && vm.VM.VmSpecSection.HardwareVersion != nil && vm.VM.VmSpecSection.HardwareVersion.Value != "" {
+		_ = d.Set("hardware_version", vm.VM.VmSpecSection.HardwareVersion.Value)
+	}
+	if vm.VM.VmSpecSection != nil && vm.VM.VmSpecSection.OsType != "" {
+		_ = d.Set("os_type", vm.VM.VmSpecSection.OsType)
+	}
```

Comment:
Not sure I understand why these checks are now needed.

CC: @vbauzysvmware

Reply:
They were needed also before. This particular code was failing when I first implemented the upgrade, and I added the check for safeguard.